### PR TITLE
Add forum

### PR DESCRIPTION
**Pull request by Fiamma-Jikukan**

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -23,58 +28,52 @@ app.use("/public", express.static(__dirname + "/public"));
...
         }
-        // find user base on the session
-        const currentUser = await User.findById(currentSession.user);
-        if (!currentUser) {
-            res.redirect('/')
-            return;
-        }
-        // if it got here it means the user is authenticated.
-        res.render("profile", currentUser)
+        const currentUser = await User.findById(req.session.user);
```
Comment (MadaraUchiha) [inline server.js]: Supposedly, you already did this work, in your session middleware, when you verified the user exists. With just a bit more code you can stick the user onto `req.user` or something, and then you don't even need to do the user lookup (again) for every call.

File: routes/login.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+const router = express.Router();
+
+router.get("/", async (req, res) => {
+    try {
+        //check if there's a session. If not, render the loging main page.
+        if (!req.session) {
+            res.render('login');
+            return;
+        }
```
Comment (MadaraUchiha) [inline routes/login.js]: Nitpick: Usually, what you do is get the "not good" paths out of the way with ifs and early returns, then do what the "happy path" is after you got all the conditions out of the way.

In this case, you would check if there's a session/user (which is unusual for the /login route) and redirect+return, then render in the last branch without ifs.

File: server.js [hunk 3]
```diff
@@ -4,13 +4,18 @@ const session = require("express-session")
...
-const bcrypt = require('bcrypt');
+const ObjectId = require('mongoose').Types.ObjectId;
 const { Schema,
     userSchema,
     User,
     connect,
     sessionSchema,
-    Session } = require("./users.js");
+    Session,
+    Post } = require("./database.js");
```
Comment (MadaraUchiha) [inline server.js]: I know we said it's lower priority, but it's getting really hard to follow what imports are used and what aren't. Do try to clean up unused imports before PRing

File: routes/main.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const http = require("http");
```
Comment (MadaraUchiha) [inline routes/main.js:1]: As far as I can tell, nothing is using this file.

File: routes/post.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+const router = express.Router();
+
+router.get("/", async (req, res) => {
+    try {
+        if (!req.session) {
+            res.clearCookie('session');
+            res.redirect('/');
+            return;
+        }
+        const currentUser = await User.findById(req.session.user);
```
Comment (MadaraUchiha) [inline routes/post.js]: Notice how the very first thing you do after every session check is getting the user? That would have been a clue that getting the user should be moved to a middleware, since you already have a related middleware that already fetches the user anyway, you should have moved this logic there.

File: routes/post.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        res.render('error', { "error-message": err })
+    }
+})
+
+router.get("/:id", async (req, res) => {
+    try {
+        if(!ObjectId.isValid(req.params.id)){
+            res.redirect('/');
+            return;
+        }
```
Comment (MadaraUchiha) [inline routes/post.js:27]: Good 👍

File: routes/post.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        const currentUser = await User.findById(req.session.user);
+        res.render("post", {
+            post: currentPost,
+            user: maker,
+            replies: repliesOfThisPost,
+            authenticated: true,
+            admin: currentUser.admin
+        })
```
Comment (MadaraUchiha) [inline routes/post.js]: Last one on this topic, if you had `req.user` and you didn't have to `User.findById`, you could have collapsed the `if` branches completely with something like


```suggestion
        res.render("post", {
            post: currentPost,
            user: maker,
            replies: repliesOfThisPost,
            authenticated: !!req.user,
            admin: !!req.user && req.user.admin, // or the nicer shorthand: req.user?.admin ?? false
        })
```

File: routes/post.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    } catch (err) {
+        res.redirect('/error')
```
Comment (MadaraUchiha) [inline routes/post.js]: In general, you don't want to `res.redirect` to the error page, because that creates a history entry in the browser, and when you refresh you'll always be on `/error` now. You should render the error template, instead.

File: routes/post.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        const currentUser = await User.findById(req.session.user);
+        const newPost = await Post.create({
+            user: { id: currentUser.id, name: currentUser.username },
+            title: req.body.title,
+            text: req.body.post,
+            created: new Date()
+        });
+        res.redirect(`/post/${newPost.id}`)
+    } catch (err) {
+        res.render('error', { "error-message": err })
```
Comment (MadaraUchiha) [inline routes/post.js:69]: Like you did here, see?

File: routes/post.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    }
+})
+
+router.post("/:id/reply", async (req, res) => {
+    try {
+        const currentPost = await Post.findById(req.params.id)
```
Comment (MadaraUchiha) [inline routes/post.js]: Not checking if the ID is a valid ObjectId this time?

In a real application, you'd have some sort of validation layer that makes sure you don't forget these things.

File: routes/post.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        if (!req.session) {
+            res.clearCookie('session');
+            res.redirect(`/post/${currentPost.id}`);
+            return;
+        }
+        const currentUser = await User.findById(req.session.user);
+        if (!currentUser) {
+            res.redirect(`/post/${currentPost.id}`);
+            return;
+        }
```
Comment (MadaraUchiha) [inline routes/post.js]: Supposedly, this is a redundant check. You wouldn't have a `req.session` if the user didn't exist, thanks to the check in your middleware.

File: routes/signup.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+const router = express.Router();
+
+router.get("/", async (req, res) => {
+    try {
+        //check if there's a session. If not, render the loging main page.
+        if (!req.session) {
+            res.render('signup');
+            return;
+        }
```
Comment (MadaraUchiha) [inline routes/signup.js]: Similarly here with the direction of the if, you would check the unusual situations first.

File: routes/profile.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    postSchema,
+    Post,
+    replySchema,
+    Reply } = require("../database.js");
+
+const router = express.Router();
+
+router.get("/:name", async (req, res) => {
+    try {
+        const pageUser = await User.find({ username: req.params.name });
```
Comment (MadaraUchiha) [inline routes/profile.js]: Why not `.findOne()` if all you're doing is looking at `[0]`?

File: routes/profile.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+            res.render('profile', { user: { username: "this user does not exist" } });
+            return;
+        }
+        const allPostOfUserQ = await Post.find({
+            user:
+            {
+                id: pageUser[0].id,
+                name: pageUser[0].username
+            }
+        })
```
Comment (MadaraUchiha) [inline routes/profile.js]: You don't need to search by both user ID and user name, the ID is a globally unique identifier, it's enough to search by it.

File: routes/profile.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+            return
+        }
+        const currentUser = await User.findById(req.session.user);
+        if (req.params.name !== currentUser.username) {
+            res.render(`profile`, { user: pageUser[0],posts: allPostOfUser, admin: currentUser.admin })
+            return;
+        }
+        res.render(`profile`, { user: currentUser, posts: allPostOfUser, session: req.session, admin: currentUser.admin })
+    } catch (err) {
+        res.redirect('/error')
```
Comment (MadaraUchiha) [inline routes/profile.js]: Same here about redirecting to `/error`

**Review by MadaraUchiha — COMMENTED** [review]
Looking much better! Added a few more comments, though they aren't as big as the last time.

File: routes/login.js (hunk 2 above)
Comment (Fiamma-Jikukan) [reply routes/login.js]: yeah, this is what I meant, habbit made the ! mark :(

File: routes/profile.js (hunk 14 above)
Comment (Fiamma-Jikukan) [reply routes/profile.js]: nope, because in posts the user IS an object with the userID and name. it's not finding it just by id, I tried. Also, I do want to make sure that this post was created by this user in case of someone killing his acout and then another user takes his name.

image

Comment (MadaraUchiha) [reply routes/profile.js]: The ID will not repeat between users with the same name, the name might, that's the point about an ID, it never repeats and never changes, even if you delete the user, that ID will never return.

Also, one of the cool parts about Mongo is that you are able to query deep into the object with no problems. With pure mongo you are able to issue the query `{ user: { id: 'whatever' } }` and it would work. Mongoose should allow you to do the same.

File: server.js [hunk 16]
```diff
@@ -23,58 +24,53 @@ app.use("/public", express.static(__dirname + "/public"));
...
+                user: item.user.name
+            }
+        }).reverse()
+        if (!req.user) {
+            res.render('forum', { posts: posts });
             return;
         }
-        // if it got here it means the user is authenticated.
-        res.render("profile", currentUser)
+        res.render('forum', { user: req.user, posts: posts, admin: req.user.admin });
```
Comment (MadaraUchiha) [inline server.js]:
```suggestion
        res.render('forum', { user: req.user, posts: posts, admin: req.user?.admin });
```

If there is no `req.user`, you'd just pass `undefined` for `user` which is perfect for the template. Same with `admin`. The `?.` operator means "go in only if exists, if not return `undefined`" So if `req.user` is `undefined`

File: routes/post.js [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        //         replies: repliesOfThisPost
+        //     })
+        //     return;
+        // }
+        res.render("post", {
+            post: currentPost,
+            user: maker,
+            replies: repliesOfThisPost,
+            authenticated: req.user,
+            admin: req.user.admin
```
Comment (MadaraUchiha) [inline routes/post.js]: Remember, `req.user` might be undefined here (try accessing this route without a session to see what I mean)! Use `?.` to guard against that, or alternatively `!!req.user && req.user.admin` - with `&&` the second operand will not be invoked if the first operand is false.

```suggestion
            admin: req.user?.admin
```

**Review by MadaraUchiha — COMMENTED** [review]
Looking better! A couple of last small things
